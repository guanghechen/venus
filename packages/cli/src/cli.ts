import { createProgram, mountSubCommandGenerate, mountSubCommandInit } from '.'

const program = createProgram()

// mount sub-command: generate
mountSubCommandGenerate(program)

// mount sub-command: init
mountSubCommandInit(program)

program.parse(process.argv)
