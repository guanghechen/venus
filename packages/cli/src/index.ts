import './env/constant'

export * from './command/generate'
export * from './core/command'
export * from './core/option'
export * from './core/generate/command'
export * from './core/generate/context'
export * from './core/generate/processor'
export * from './env/constant'
export * from './env/logger'
