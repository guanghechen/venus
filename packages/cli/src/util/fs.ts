import fs from 'fs-extra'

/**
 * Check whether if the given filepath exists.
 * @param filepath
 */
export function isExists(filepath: string | null): filepath is string {
  return filepath !== null && fs.existsSync(filepath)
}

/**
 * Check whether if the given filepath exists and that it is a file path.
 * @param filepath
 */
export function isFile(filepath: string | null): filepath is string {
  return isExists(filepath) && fs.statSync(filepath).isFile()
}

/**
 * Check whether if the given filepath exists and that it is a directory path.
 * @param filepath
 */
export function isDirectory(filepath: string | null): boolean {
  return isExists(filepath) && fs.statSync(filepath).isDirectory()
}

/**
 * Ensure the given filepath exists, otherwise, an exception will be thrown.
 * @param filepath
 * @param message
 */
export function ensureExists(
  filepath: string | null,
  message?: string,
): void | never {
  if (isExists(filepath)) return
  throw new ReferenceError(message ?? `Cannot find '${filepath}'.`)
}

/**
 * Ensure the given filepath exists and it is a file path.
 * Otherwise, an exception will be thrown.
 * @param filepath
 * @param message
 */
export function ensureFileExists(
  filepath: string | null,
  message?: string,
): void | never {
  ensureExists(filepath, message)
  if (fs.statSync(filepath!).isFile()) return
  throw new ReferenceError(message ?? `${filepath} is not a file path.`)
}

/**
 * Ensure the given filepath exists and it is a directory path.
 * Otherwise, an exception will be thrown.
 * @param filepath
 * @param message
 */
export function ensureDirectoryExists(
  filepath: string | null,
  message?: string,
): void | never {
  ensureExists(filepath, message)
  if (fs.statSync(filepath!).isDirectory()) return
  throw new ReferenceError(message ?? `${filepath} is not a directory path.`)
}
