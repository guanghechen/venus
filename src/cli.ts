import { createProgram, mountSubCommandGenerate } from '.'

const program = createProgram()

// mount sub-command: generate
mountSubCommandGenerate(program)

program.parse(process.argv)
