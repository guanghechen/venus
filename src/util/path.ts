import { isNonBlankString } from '@guanghechen/option-helper'
import path from 'path'

/**
 * Check if filepath1 under the filepath2.
 *
 * @param filepath1
 * @param filepath2
 * @see https://stackoverflow.com/a/45242825/15760674
 */
export function isUnderThePath(filepath1: string, filepath2: string): boolean {
  const p = path.relative(filepath2, filepath1)
  return isNonBlankString(p) && !p.startsWith('..') && !path.isAbsolute(p)
}

/**
 * Calc the path of filepath relative to cwd.
 *
 * @param filepath          target filepath
 * @param cwd               current workspace directory
 * @param rootDirectory     base directory
 */
export function relativePath(
  filepath: string,
  cwd: string,
  rootDirectory?: string,
): string {
  const absoluteFilepath = path.resolve(cwd, filepath)
  if (rootDirectory != null) {
    if (!isUnderThePath(absoluteFilepath, rootDirectory)) {
      return absoluteFilepath
    }
  }
  return path.relative(cwd, filepath)
}
