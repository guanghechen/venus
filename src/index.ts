import loadCommand from '@/command'
import { doneWithClose } from '@/util/cli'
import { registerCommanderOptions } from '@barusu/chalk-logger'
import program from 'commander'
import manifest from '../package.json'

// 改造 Commander 的 action 方法，使其在执行成功后退出程序（为了退出控制台输入）
{
  const fn = program.Command.prototype as any
  const action = fn.action
  fn.action = function (f: (...args: any[]) => Promise<void>) {
    return action.call(this, doneWithClose(f))
  }
}

program.version(manifest.version)

registerCommanderOptions(program)

loadCommand(program)

program.parse(process.argv)
