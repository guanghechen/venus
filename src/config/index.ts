import { isFile } from '@/util/fs'
import fs from 'fs-extra'
import yaml from 'js-yaml'
import path from 'path'
import { DefaultCreateConfig } from './create'
import { DefaultGenerateConfig } from './generate'
import type { RawDefaultCreateConfig } from './create'
import type { RawDefaultGenerateConfig } from './generate'

const absoluteLocaleConfigPath = path.join(__dirname, 'config.yml')
const localConfigContent = fs.readFileSync(absoluteLocaleConfigPath, 'utf-8')
const localRawConfig = yaml.load(localConfigContent) as RawPartialConfig

/**
 * 配置文件的类型
 */
export interface RawPartialConfig {
  generate: RawDefaultGenerateConfig
  create: RawDefaultCreateConfig
}

/**
 * 获取外部的配置文件
 *
 * @param projectDirectory
 * @param configPath
 */
export const getPartialRawConfig = async (
  projectDirectory: string,
  configPath?: string,
): Promise<RawPartialConfig | undefined> => {
  const absoluteConfigPath = path.resolve(
    projectDirectory,
    configPath != null ? configPath : 'venus.config.yml',
  )
  if (isFile(absoluteConfigPath)) {
    const configContent = fs.readFileSync(absoluteConfigPath, 'utf-8')
    return yaml.load(configContent) as unknown as RawPartialConfig
  }
}

/**
 * 获取子命令 'generate' 的默认选项
 */
export const getDefaultGenerateConfig = (
  partialRawConfig?: RawDefaultGenerateConfig,
): DefaultGenerateConfig => {
  const { generate } = localRawConfig
  return new DefaultGenerateConfig(generate, partialRawConfig)
}

/**
 * 获取子命令 'create/new' 的默认选项
 */
export const getDefaultCreateConfig = (
  partialRawConfig?: RawDefaultCreateConfig,
): DefaultCreateConfig => {
  const { create } = localRawConfig
  return new DefaultCreateConfig(create, partialRawConfig)
}
