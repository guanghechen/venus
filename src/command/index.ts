import path from 'path'
import commander from 'commander'
import { projectConfig } from '@/config/immutable'
import { getPartialRawConfig, RawPartialConfig } from '@/config'
import { ensureFileExist, findNearestTarget } from '@/util/fs-util'
import { coverString } from '@/util/option-util'
import loadGenerateCommand from './generate'
import loadCreateCommand from './create'
import loadRegisterCommand from './register'


export interface GlobalConfig {
  readonly executeDirectory: string
  readonly projectRootDirectory: string
  readonly definitionPhase: string
  readonly encoding: string
  readonly cmakeLists: {
    readonly encoding: string
    readonly filepath: string
  }
}


export default (program: commander.Command) => {
  program
    .option('--cmake-lists <cmakeLists-name>', 'index cmakeLists name. (default is CMakeLists.txt)')
    .option('--encoding <encoding>', 'index encoding of all files.')
    .option('--config-path <config-path>', 'index config path, related with the project root directory where the CMakeLists.txt exists.')
    .option('--no-config', 'don\'t use config.')

  // 挂载 generate 子命令
  loadGenerateCommand(program, getRawPartialConfig('generate'), getGlobalConfig)

  // 挂载 create/new 子命令
  loadCreateCommand(program, getRawPartialConfig('create'), getGlobalConfig)

  // 挂载 register 子命令
  loadRegisterCommand(program, getRawPartialConfig('register'), getGlobalConfig)

  /**
   * 获取外部配置文件
   * @param key
   */
  function getRawPartialConfig(key: 'generate' | 'create' | 'register') {
    return async (specifiedProjectLocatedPath?: string): Promise<RawPartialConfig|any|undefined> => {
      const globalConfig: GlobalConfig = await getGlobalConfig(specifiedProjectLocatedPath)
      if (program.config === false) return
      const partialRawConfig: RawPartialConfig | undefined = await getPartialRawConfig(globalConfig.projectRootDirectory, program.configPath)
      if (partialRawConfig == null) return
      return partialRawConfig[key]
    }
  }

  /**
   * 生成全局选项
   *
   * @param specifiedProjectLocatedPath   该值为绝对路径时才生效，从指定的地方搜索 CMakeLists.txt
   */
  async function getGlobalConfig(specifiedProjectLocatedPath?: string): Promise<GlobalConfig> {
    const { definitionPhase, cmakeLists, encoding } = projectConfig
    const cmakeListsFileName = coverString(cmakeLists.filename, program.cmakeLists)
    const cmakeListsEncoding = coverString(cmakeLists.encoding, program.encoding)

    const executeDirectory = path.resolve()
    if (specifiedProjectLocatedPath != null) specifiedProjectLocatedPath = path.resolve(executeDirectory, specifiedProjectLocatedPath)
    const absoluteCmakeListsPath = specifiedProjectLocatedPath != null
      ? await findNearestTarget(specifiedProjectLocatedPath, cmakeListsFileName)
      : await findNearestTarget(executeDirectory, cmakeListsFileName)

    // 确保 CMakeLists.txt 存在
    await ensureFileExist(absoluteCmakeListsPath, `'${cmakeListsFileName}' is not found.`)

    const projectRootDirectory: string = path.dirname(absoluteCmakeListsPath!)
    return {
      executeDirectory,
      projectRootDirectory,
      definitionPhase,
      encoding,
      cmakeLists: {
        filepath: absoluteCmakeListsPath!,
        encoding: cmakeListsEncoding,
      },
    }
  }
}
