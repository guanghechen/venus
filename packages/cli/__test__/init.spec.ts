import { fileSnapshot } from '@guanghechen/helper-jest'
import { absoluteOfWorkspace } from '@guanghechen/helper-path'
import globby from 'globby'
import fs from 'node:fs'
import path from 'node:path'
import url from 'node:url'
import { COMMAND_NAME, createProgram, execSubCommandInit } from '../src'

const __dirname = path.dirname(url.fileURLToPath(import.meta.url))

describe('init', function () {
  // clear output directory before run test
  const outputWorkspace = absoluteOfWorkspace(__dirname, '__tmp__/simple')
  if (fs.existsSync(outputWorkspace))
    fs.rmSync(outputWorkspace, { recursive: true })

  test(
    'simple',
    async function () {
      const program = createProgram()
      await execSubCommandInit(program, [
        '',
        COMMAND_NAME,
        'init',
        outputWorkspace,
        '--plop-bypass=simple', // select template
        '--plop-bypass=' + path.basename(outputWorkspace), // package name
        '--plop-bypass=utf-8', // encoding
        '--plop-bypass=verbose', // log level
        '--plop-bypass=skip', // skip 'yarn install'
        '--plop-bypass=no', // skip 'git commit'
        '--log-level=warn',
      ])

      // write the outputs to snapshots
      const files = (
        await globby.globby(['*', '**/*'], {
          cwd: outputWorkspace,
          onlyFiles: true,
          expandDirectories: false,
        })
      ).sort()

      fileSnapshot(outputWorkspace, files)
    },
    1000 * 30,
  )
})
