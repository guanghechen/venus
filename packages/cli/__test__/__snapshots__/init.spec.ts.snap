// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`init simple: CMakeLists.txt 1`] = `
"cmake_minimum_required(VERSION 3.15)
project(acm)
set(CMAKE_CXX_STANDARD 11)

include_directories(src)

add_compile_definitions(GUANGHECHEN_LOCAL_CMAKE)
add_compile_definitions(_GLIBCXX_DEBUG)

# fake problems
file(GLOB_RECURSE OJ_SOURCES oj/fake/*.cpp)
foreach(SOURCE_PATH \${OJ_SOURCES})
  get_filename_component(EXECUTABLE_NAME \${SOURCE_PATH} NAME_WE)
  add_executable(\\"FAKE_\${EXECUTABLE_NAME}\\" \${SOURCE_PATH})
endforeach(SOURCE_PATH \${OJ_SOURCES})
"
`;

exports[`init simple: oj/fake/01.cpp 1`] = `
"#include <@guanghechen/algorithm/match-first.hpp>
#include <cstdio>

int main() {
  int data[5] = { 1, 2, 3, 3, 4 };
  int result = matchFirst(
    0,
    4,
    [&data](int x) -> bool { return data[x] >= 3; }
  );
  printf(\\"result: %d\\\\n\\", result);
  return 0;
}
"
`;

exports[`init simple: package.json 1`] = `
"{
  \\"name\\": \\"@guanghechen/acm\\",
  \\"version\\": \\"0.0.0\\",
  \\"private\\": \\"true\\",
  \\"scripts\\": {
    \\":upgrade\\": \\"yarn upgrade-interactive --latest && run-s clean format\\",
    \\"clean\\": \\"yarn\\",
    \\"cpp:generate\\": \\"venus generate\\",
    \\"cpp:run\\": \\"node script/run.js\\",
    \\"format\\": \\"run-s format:lint:fix format:prettier format:cpp\\",
    \\"format:lint:fix\\": \\"eslint . --fix\\",
    \\"format:cpp\\": \\"git-clang-format\\",
    \\"format:prettier\\": \\"prettier --write .\\",
    \\"postinstall\\": \\"run-p postinstall:**\\",
    \\"postinstall:main\\": \\"yarn update:lock\\",
    \\"postinstall:dev\\": \\"is-ci || husky install\\",
    \\"update:lock\\": \\"yarn-deduplicate\\"
  },
  \\"devDependencies\\": {
    \\"@babel/core\\": \\"^7.14.3\\",
    \\"@guanghechen/eslint-config\\": \\"^1.5.1\\",
    \\"@guanghechen/eslint-config-ts\\": \\"^1.5.1\\",
    \\"@types/node\\": \\"^15.3.1\\",
    \\"chalk\\": \\"^4.1.1\\",
    \\"clang-format\\": \\"^1.5.0\\",
    \\"eslint\\": \\"^7.27.0\\",
    \\"eslint-config-prettier\\": \\"^8.3.0\\",
    \\"eslint-plugin-prettier\\": \\"^3.4.0\\",
    \\"fs-extra\\": \\"^10.0.0\\",
    \\"husky\\": \\"6.0.0\\",
    \\"is-ci\\": \\"3.0.0\\",
    \\"lint-staged\\": \\"11.0.0\\",
    \\"minimist\\": \\"^1.2.5\\",
    \\"npm-run-all\\": \\"^4.1.5\\",
    \\"prettier\\": \\"^2.3.0\\",
    \\"ts-node\\": \\"^9.1.1\\",
    \\"tsconfig-paths\\": \\"^3.9.0\\",
    \\"typescript\\": \\"^4.2.4\\",
    \\"venus-acm\\": \\"2.x.x\\",
    \\"yarn-deduplicate\\": \\"3.1.0\\"
  },
  \\"lint-staged\\": {
    \\"*.{js,jsx,ts,tsx}\\": [
      \\"eslint --fix\\",
      \\"prettier --write\\"
    ],
    \\"*.{json,eslintrc}\\": [
      \\"prettier --write\\"
    ],
    \\"*.{cpp,hpp}\\": [
      \\"git-clang-format\\"
    ]
  },
  \\"venus-acm\\": {
    \\"generate\\": {
      \\"includes\\": [],
      \\"force\\": true,
      \\"copy\\": true,
      \\"uglify\\": false,
      \\"removeComments\\": true,
      \\"removeSpaces\\": false,
      \\"removeFreopen\\": true,
      \\"removeAssert\\": true
    }
  }
}
"
`;

exports[`init simple: script/run.js 1`] = `
"#! /usr/bin/env node

const chalk = require('chalk')
const { spawn } = require('child_process')
const fs = require('fs-extra')
const minimist = require('minimist')
const path = require('path')

const workspaceRoot = path.join(__dirname, '..')
const sourceRoot = path.join(workspaceRoot, 'src')

/**
 * Print usage information strings.
 */
function usage() {
  const [arg1, arg2] = process.argv
  console.log(\`\${arg1} \${arg2} <cpp source file>\`)
}

/**
 * Compile and run a cpp file.
 *
 * @param {string} filename
 * @param {{ verbose: boolean }} options
 */
function compileAndRun(filepath, options) {
  const { dir, base: filename, name, ext } = path.parse(filepath)
  const executableFilepath = path.join(dir, name)
  const problemId = /^\\\\d+/.test(name) ? /^(\\\\d+)/.exec(name)[1] : null

  if (ext !== '.cpp') {
    console.log(chalk.yellow('Cannot handle \`ext\` type file.'))
    return
  }

  const command = [
    \`cd \${dir}\`,
    \`g++ -std=c++11 -I '\${sourceRoot}' '\${filename}' -o '\${name}'\`,
    \`./\${name}\`,
  ].join(' && ')

  if (options.verbose) console.log(chalk.green(command))
  const child = spawn(command, { shell: true })

  const dataFilepath = path.join(dir, 'data', \`\${problemId}.in\`)
  if (fs.existsSync(dataFilepath)) {
    const stream = fs.createReadStream(dataFilepath, 'utf-8')
    stream.pipe(child.stdin)
  } else {
    process.stdin.pipe(child.stdin)
  }

  child.stdout.pipe(process.stdout)
  child.stderr.pipe(process.stdout)

  // handling CTRL + C
  child.on('exit', () => void fs.unlinkSync(executableFilepath))
}

function run() {
  const [, , ...argv] = process.argv
  const args = minimist(argv)

  const target = args._[0]
  if (target === null) {
    usage()
    return
  }

  compileAndRun(path.resolve(process.cwd(), target), {
    verbose: Boolean(args.verbose),
  })
}

run()
"
`;

exports[`init simple: src/@guanghechen/algorithm/match-first.hpp 1`] = `
"#include <functional>

/**
 * Find the first matched position.
 */
int matchFirst(int lft, int rht, std::function<bool(int)> verify) {
  int rightBound = rht;
  while (lft < rht) {
    int mid = (lft + rht) >> 1;
    if (verify(mid))
      rht = mid;
    else
      lft = mid + 1;
  }
  return rht >= rightBound ? -1 : rht;
}
"
`;
